File: reth/pyreth/examples/chain_query/balance_changes.py
#!/usr/bin/env python3
"""
Example: Track balance changes between blocks
"""

from pyreth import chain_query as pyreth_chain_query
from eth_token.erc20_token.pools.addresses import require_checksum_address, same_address

def format_wei(wei_str):
    """Format wei to ETH with 4 decimals"""
    return f"{int(wei_str) / 10**18:.4f}"

def format_token_amount(amount_str, decimals=6):
    """Format token amount with proper decimals"""
    return f"{int(amount_str) / 10**decimals:,.2f}"

def main():
    # Create PyReth instance
    query = pyreth_chain_query()
    
    print("=== Chain Query: Balance Changes ===\n")
    
    # Get latest block
    latest_block = query.get_latest_block()
    print(f"Latest block: {latest_block}")
    
    # Set block range (last 100 blocks)
    from_block = latest_block - 100
    to_block = latest_block
    
    print(f"Analyzing blocks {from_block} to {to_block}")
    print()
    
    # Test addresses
    addresses = [
        ("Vitalik", "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"),
        ("Binance", "0xF977814e90dA44bFA03b6295A0616a897441aceC"),
    ]
    
    # Token addresses
    usdc = require_checksum_address("0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48")
    usdt = require_checksum_address("0xdAC17F958D2ee523a2206206994597C13D831ec7")
    
    tokens = [usdc, usdt]
    
    for name, addr in addresses:
        print(f"Balance changes for {name}")
        print("-" * 40)
        
        try:
            # Get balance changes
            changes = query.get_balance_changes(addr, tokens, from_block, to_block)
            
            # Display ETH changes
            print(f"ETH Changes:")
            print(f"  Before: {format_wei(changes.eth_change.before)} ETH")
            print(f"  After:  {format_wei(changes.eth_change.after)} ETH")
            
            if changes.eth_change.is_increase:
                print(f"  Change: +{format_wei(changes.eth_change.difference)} ETH ↑")
            else:
                # For decrease, the difference is already negative in the string
                diff_wei = changes.eth_change.difference.lstrip('-')
                print(f"  Change: -{format_wei(diff_wei)} ETH ↓")
            
            # Display token changes (if any)
            token_changes = changes.token_changes()
            if token_changes:
                print(f"\nToken Changes:")
                for token_addr, change_data in token_changes.items():
                    token_name = "USDC" if same_address(token_addr, usdc) else "USDT"
                    print(f"  {token_name}: {change_data}")
            
        except Exception as e:
            print(f"  Error getting changes: {e}")
        
        print()
    
    # Example: Find significant balance changes
    print("Finding Significant Changes")
    print("-" * 40)
    
    # Check multiple addresses for significant ETH movements
    exchange_addresses = [
        ("Coinbase", "0xA9D1e08C7793af67e9d92fe308d5697FB81d3E43"),
        ("Kraken", "0x53d284357ec70cE289D6D64134DfAc8E511c8a3D"),
    ]
    
    print("Checking exchange balance changes...")
    for name, addr in exchange_addresses:
        try:
            changes = query.get_balance_changes(addr, [], from_block, to_block)
            
            before_eth = int(changes.eth_change.before) / 10**18
            after_eth = int(changes.eth_change.after) / 10**18
            diff_eth = after_eth - before_eth
            
            if abs(diff_eth) > 0.1:  # Significant if > 0.1 ETH
                symbol = "↑" if diff_eth > 0 else "↓"
                print(f"  {name}: {diff_eth:+.4f} ETH {symbol}")
        except Exception as e:
            print(f"  {name}: Error - {e}")
    
    print()
    
    # Example: Portfolio tracking
    print("Portfolio Tracking")
    print("-" * 40)
    
    vitalik_addr = addresses[0][1]
    
    # Get portfolio at two different blocks
    try:
        portfolio_before = query.get_portfolio(vitalik_addr, tokens, from_block)
        portfolio_after = query.get_portfolio(vitalik_addr, tokens, to_block)
        
        print(f"Vitalik's portfolio changes:")
        print(f"  Block {from_block}: {format_wei(portfolio_before.eth_balance)} ETH")
        print(f"  Block {to_block}: {format_wei(portfolio_after.eth_balance)} ETH")
        
        eth_diff = (int(portfolio_after.eth_balance) - int(portfolio_before.eth_balance)) / 10**18
        if eth_diff != 0:
            print(f"  Net change: {eth_diff:+.4f} ETH")
        else:
            print(f"  No change in ETH balance")
            
    except Exception as e:
        print(f"Error tracking portfolio: {e}")
    
    print("\n✅ Balance change tracking complete!")

if __name__ == "__main__":
    main()

File: reth_chain_query/examples/address_state/balance_changes.rs
use alloy_primitives::{utils::format_ether, utils::format_units, Address, U256};
use reth_chain_query::provider::BalanceDiff;
/// Track balance changes between blocks
///
/// This example demonstrates:
/// 1. get_balance_changes - Track ETH and token changes
/// 2. Identifying deposits and withdrawals
/// 3. Multi-token change tracking
/// 4. Exchange flow monitoring
///
/// Run with: cargo run --example balance_changes
use reth_chain_query::{Result, RethQueryProvider};
use std::str::FromStr;

// Tokens to track
const TOKENS: &[(&str, &str, u8)] = &[
    ("USDC", "A0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", 6),
    ("USDT", "dAC17F958D2ee523a2206206994597C13D831ec7", 6),
    ("WETH", "C02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", 18),
];

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Balance Change Tracking ===\n");

    let reth_datadir = tx_simulator::config::repo::reth_datadir()?;
    let provider = RethQueryProvider::new(&reth_datadir)?;
    let latest_block = provider.get_latest_block()?;

    // === 1. Single Address Balance Changes ===
    println!("1. Address Balance Changes");
    println!("{}", "-".repeat(60));

    // Track Binance hot wallet changes over recent blocks
    let binance = Address::from_str("F977814e90dA44bFA03b6295A0616a897441aceC")?;
    let from_block = latest_block - 1000; // Last ~3.3 hours
    let to_block = latest_block;

    println!(
        "Tracking Binance changes from block {} to {}",
        from_block, to_block
    );
    println!(
        "Time period: ~{:.1} hours\n",
        (to_block - from_block) as f64 * 12.0 / 3600.0
    );

    let token_addresses: Vec<Address> = TOKENS
        .iter()
        .map(|(_, addr, _)| Address::from_str(addr).unwrap())
        .collect();

    let changes = provider
        .calculate_eth_and_token_balance_diff_between_blocks(
            binance,
            token_addresses.clone(),
            from_block,
            to_block,
        )
        .await?;

    // Display ETH changes
    println!("ETH Balance Changes:");
    println!("  Before: {} ETH", format_ether(changes.eth_change.before));
    println!("  After:  {} ETH", format_ether(changes.eth_change.after));

    match &changes.eth_change.difference {
        BalanceDiff::Increase(amount) => {
            println!("  Change: +{} ETH ✅", format_ether(*amount));
        }
        BalanceDiff::Decrease(amount) => {
            println!("  Change: -{} ETH ❌", format_ether(*amount));
        }
    }

    println!();

    // Display token changes
    if !changes.token_changes.is_empty() {
        println!("Token Balance Changes:");

        for (token_addr, change) in &changes.token_changes {
            if let Some((symbol, _, decimals)) = TOKENS
                .iter()
                .find(|(_, addr, _)| Address::from_str(addr).unwrap() == *token_addr)
            {
                println!("\n  {}:", symbol);
                println!("    Before: {}", format_units(change.before, *decimals)?);
                println!("    After:  {}", format_units(change.after, *decimals)?);

                match &change.difference {
                    BalanceDiff::Increase(amount) => {
                        println!("    Change: +{} ✅", format_units(*amount, *decimals)?);
                    }
                    BalanceDiff::Decrease(amount) => {
                        println!("    Change: -{} ❌", format_units(*amount, *decimals)?);
                    }
                }
            }
        }
    }

    println!();

    // === 2. Multiple Exchange Comparison ===
    println!("2. Exchange Flow Comparison");
    println!("{}", "-".repeat(60));

    let exchanges = vec![
        ("Binance", "F977814e90dA44bFA03b6295A0616a897441aceC"),
        ("Coinbase", "A9D1e08C7793af67e9d92fe308d5697FB81d3E43"),
        ("Kraken", "53d284357ec70cE289D6D64134DfAc8E511c8a3D"),
    ];

    println!("Comparing ETH flows over last 100 blocks (~20 minutes):\n");
    let recent_from = latest_block - 100;

    for (name, addr_str) in &exchanges {
        let address = Address::from_str(addr_str)?;

        let changes = provider
            .calculate_eth_and_token_balance_diff_between_blocks(
                address,
                vec![], // Only ETH for this comparison
                recent_from,
                latest_block,
            )
            .await?;

        print!("{:10} ", format!("{}:", name));

        match &changes.eth_change.difference {
            BalanceDiff::Increase(amount) => {
                println!("+{:>12} ETH ⬆️", format_ether(*amount));
            }
            BalanceDiff::Decrease(amount) => {
                println!("-{:>12} ETH ⬇️", format_ether(*amount));
            }
        }
    }

    println!();

    // === 3. Historical Balance Evolution ===
    println!("3. Balance Evolution Over Time");
    println!("{}", "-".repeat(60));

    let address = Address::from_str("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045")?; // Vitalik

    println!("Tracking Vitalik's ETH balance evolution:\n");

    // Define checkpoints (roughly every 1000 blocks)
    let checkpoints = vec![
        latest_block - 3000,
        latest_block - 2000,
        latest_block - 1000,
        latest_block,
    ];

    let mut previous_balance = None;

    for &block in &checkpoints {
        let account = provider.get_account(address, Some(block)).await?;
        let balance = account.balance;

        print!("Block {:8}: {:>15} ETH", block, format_ether(balance));

        if let Some(prev) = previous_balance {
            if balance > prev {
                let diff = balance - prev;
                print!(" (+{} ETH)", format_ether(diff));
            } else if balance < prev {
                let diff = prev - balance;
                print!(" (-{} ETH)", format_ether(diff));
            } else {
                print!(" (unchanged)");
            }
        }

        println!();
        previous_balance = Some(balance);
    }

    println!();

    // === 4. Net Flow Analysis ===
    println!("4. Net Flow Summary");
    println!("{}", "-".repeat(60));

    println!("Calculating net flows for major exchanges (last 500 blocks):\n");

    let flow_from = latest_block - 500;
    let mut total_inflow = U256::ZERO;
    let mut total_outflow = U256::ZERO;

    for (name, addr_str) in &exchanges {
        let address = Address::from_str(addr_str)?;

        let changes = provider
            .calculate_eth_and_token_balance_diff_between_blocks(
                address,
                vec![],
                flow_from,
                latest_block,
            )
            .await?;

        match &changes.eth_change.difference {
            BalanceDiff::Increase(amount) => {
                total_inflow = total_inflow + *amount;
                println!("{:10} Net inflow:  +{} ETH", name, format_ether(*amount));
            }
            BalanceDiff::Decrease(amount) => {
                total_outflow = total_outflow + *amount;
                println!("{:10} Net outflow: -{} ETH", name, format_ether(*amount));
            }
        }
    }

    println!("\nAggregate Flow:");
    println!("  Total Inflow:  {} ETH", format_ether(total_inflow));
    println!("  Total Outflow: {} ETH", format_ether(total_outflow));

    if total_inflow > total_outflow {
        println!(
            "  Net Result:    +{} ETH (Accumulation)",
            format_ether(total_inflow - total_outflow)
        );
    } else {
        println!(
            "  Net Result:    -{} ETH (Distribution)",
            format_ether(total_outflow - total_inflow)
        );
    }

    println!("\n✅ Balance change tracking complete!");
    println!("\nKey Insights:");
    println!("- Track balance changes between any two blocks");
    println!("- Monitor exchange inflows and outflows");
    println!("- Analyze token movements alongside ETH");
    println!("- Useful for whale watching and flow analysis");

    Ok(())
}
